File: Lasso - Ridge Regression/ToyotaCorolla_Lasso_Ridge.R
library(readr)
corolla <- read_csv(file.choose())
View(corolla)

corolla <- corolla[c("Price","Age_08_04","KM","HP","cc","Doors","Gears","Quarterly_Tax","Weight")]
View(corolla)

# Data Preparation & EDA

# checking for missing values

sum(is.na(corolla)) # No missing values

# 6 - point Summary

summary(corolla)
colnames(corolla)

attach(corolla)

library(dplyr)
summarise_if(corolla , is.numeric, sd) # Standard Deviation
summarise_if(corolla , is.numeric, max) - summarise_if(corolla , is.numeric, min) # Range

# Skewness
#install.packages("moments")
library(moments)
skewness(corolla)

#Kurtosis
kurtosis(corolla)

#Graphical Representation

#Distribution
dens <- apply(corolla , 2 , density)
plot(NA, xlim=range(sapply(dens, "[", "x")), ylim=range(sapply(dens, "[", "y")))
mapply(lines, dens, col=1:length(dens))
legend("topright", legend=names(dens), fill=1:length(dens))

#Box Plot
boxplot(corolla) # Outliers exists for Price, KM, CC

# Pair plot
pairs(corolla)

# Correlation Matrix
panel.cor <- function(x, y, digits=2, prefix="", cex.cor)
{
  #usr <- par("usr"); on.exit(par(usr))
  par(usr = c(0, 1, 0, 1))
  r = (cor(x, y))
  txt <- format(c(r, 0.123456789), digits=digits)[1]
  txt <- paste(prefix, txt, sep="")
  if(missing(cex.cor)) cex <- 0.4/strwidth(txt)
  text(0.5, 0.5, txt, cex = cex)
}
pairs(corolla, upper.panel=panel.cor,main="Scatter Plot Matrix with Correlation Coefficients")


#Linear Model
model1 <- lm(Price ~ . , data = corolla)
summary(model1)

# Output Summary:
# 1. p-values of most of the coeffs are less than 0.05 but cc and Doors are insignificant
# R squared adj and R squared are closeby
# R squared value is 0.864 - Determination is strong 


# Diagnostic Plots
#install.packages("car")
library(car)
plot(model1)


# Deletion Diagnostics for identifying influential variable
influence.measures(model1)
influenceIndexPlot(model1, id.n=3) 
influencePlot(model1, id.n=3) # Points 81, 222 & 961 seems to be outliers


# Converting the data into format in which model accepts 
x <- model.matrix(Price~.-1,data=corolla)
y <- corolla$Price

library(glmnet)

# setting lamda as 10^10 till 10^-2
lambda <- 10^seq(10, -2, length = 50)

# Ridge Regression
ridge_reg <- glmnet(x,y,alpha=0,lambda=lambda)
summary(ridge_reg)

# Coefficients vs Lambda
plot(ridge_reg,xvar="lambda",label=T)

# ridge regression coefficients, stored in a matrix 
dim(coef(ridge_reg))
plot(ridge_reg)

# Cross Validation in Ridge Regression
cv_ridge = cv.glmnet(x, y, alpha = 0) 

# Select lamda that minimizes training MSE
bestlam = cv_ridge$lambda.min  
bestlam

# Draw plot of training MSE as a function of lambda
plot(cv_ridge) 

ridge_pred1 = predict(ridge_reg, s = bestlam, newx = x)
sqrt(mean((ridge_pred1 - y)^2)) # RMSE is 1384.5

ridge_pred1 = predict(ridge_reg, s = 0, newx = x)
sqrt(mean((ridge_pred1 - y)^2)) #RMSE is 1338.2 indicates basic regression is better than lasso regression


# LASSO Regression
lasso_mod = glmnet(x,y, alpha = 1, lambda = lambda)

plot(lasso_mod)    # Draw plot of coefficients

cv_lasso = cv.glmnet(x, y, alpha = 1) # Fit lasso model on training data

plot(cv_lasso) # Draw plot of training MSE as a function of lambda

bestlam_lasso = cv_lasso$lambda.min # Select lamda that minimizes training MSE
bestlam_lasso
# Use best lambda to predict test data
lasso_pred = predict(lasso_mod, s = bestlam_lasso, newx = x)

sqrt(mean((lasso_pred - y)^2)) # RMSE is 8981.8 which is is higher than basic multiple linear regression

# Display coefficients using lambda chosen by CV
lasso_coef = predict(lasso_mod, type = "coefficients", s = bestlam)[1:5,] 
lasso_coef


#######
# Partitioning Data into training set and testing set
# Spliting data into train and test
n=nrow(corolla)
n1=n*0.7
n2=n-n1
train=sample(1:n,n1)
test = corolla[-train,]

pred=predict(model1,newdat=test)
actual=test$Price
error=actual-pred

test.rmse=sqrt(mean(error**2))
test.rmse

train.rmse = sqrt(mean(model1$residuals**2))
train.rmse 

# Train error and Test error are similar So the model is having a good fit

File: Lasso - Ridge Regression/ToyotaCorolla_Lasso_Ridge.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

corolla = pd.read_csv('C:/Users/lnram/Desktop/ExcelR/Module 7 - Multiple Linear Regresion/Assignments/ToyotaCorolla.csv' , encoding= 'unicode_escape')

corolla = corolla[["Price","Age_08_04","KM","HP","cc","Doors","Gears","Quarterly_Tax","Weight"]]

# Check for missing values
corolla.isna().sum()

# Basic Statistics
corolla.describe()

corolla.dtypes

#Skeweness
corolla.skew() # CC value is skewed positively

#Kurtosis
corolla.kurt() # CC value is leptokurtic

#Graphical Representation

len(corolla.columns)

i = 1
for cols in corolla.columns:
    plt.subplot(3,3,i)
    sns.distplot(corolla[cols] , hist = False , label = cols)    
    i = i+1

# Positively Skewed - Price, KM, HP, CC, Gears, Weight
# Negatively Skewed - Age
# ultiple Disributions - Doors, Quarterly_Tax


corolla.boxplot(grid = False)

# Outliers exist is all features except for Doors

# Model Building - Multiple Linear Regression
# Steps - Build model, Score, Predicting, RMSE, Model Assumptions

from sklearn.linear_model import LinearRegression
LR1 = LinearRegression()

x = corolla.iloc[:,1:]
y = corolla.iloc[:,0]


LR1.fit(x , y)
# Getting coefficients of variables               
LR1.coef_
LR1.intercept_

# Getting Adjusted R squared value
LR1.score(x,y)

pred1 = LR1.predict(x)

# Rmse value
np.sqrt(np.mean((pred1-y)**2))

# Validating Model Assumptions
# Residuals Vs Fitted Values
plt.scatter(x = pred1 , y = y - pred1);plt.xlabel("Fitted");plt.ylabel("Residuals")
# Checking normal distribution 
plt.hist(pred1-y) # Normally distributed but skewed

# Predicted Vs Actual
plt.scatter(x = pred1 , y = y);plt.xlabel("Predicted");plt.ylabel("Actual")
plt.bar(height = pd.Series(LR1.coef_),x = x.columns) #clearly the coeff magnitudes vary by large extent

# Running Ridge Regression
from sklearn.linear_model import Ridge
RM1 = Ridge(alpha = 0.4,normalize=True)
RM1.fit(x,y)
# Coefficient values for all the independent variables
RM1.coef_
RM1.intercept_
plt.bar(height = pd.Series(RM1.coef_),x=pd.Series(x.columns))

# Getting Adjusted R squared value
RM1.score(x,y)

pred1 = RM1.predict(x)

# Rmse value
np.sqrt(np.mean((pred1-y)**2))

# Validating Model Assumptions
# Residuals Vs Fitted Values
plt.scatter(x = pred1 , y = y - pred1);plt.xlabel("Fitted");plt.ylabel("Residuals")
# Checking normal distribution 
plt.hist(pred1-y) # Normally distributed but skewed

# Predicted Vs Actual
plt.scatter(x = pred1 , y = y);plt.xlabel("Predicted");plt.ylabel("Actual")
plt.bar(height = pd.Series(LR1.coef_),x = x.columns) #clearly the coeff magnitudes vary by large extent

# Running a Ridge Regressor of set of alpha values and observing how the R-Squared
rmse = []
R_sqrd = []
alphas = np.arange(0,100,0.05)
for i in alphas:
    RM = Ridge(alpha = i,normalize=True)
    RM.fit(x , y)
    R_sqrd.append(RM.score(x , y))
    rmse.append(np.sqrt(np.mean((RM.predict(x) - y)**2)))
    
# Plotting RMSE,R_Squared values with respect to alpha values

# Alpha vs R_Squared values
plt.scatter(x=alphas,y=R_sqrd);plt.xlabel("alpha");plt.ylabel("R_Squared")

# Alpha vs RMSE
plt.scatter(x=alphas,y=rmse);plt.xlabel("alpha");plt.ylabel("rmse")

# We got minimum R_Squared value at small alpha values 

# Running Lasso Regression
from sklearn.linear_model import Lasso
LM1 = Lasso(alpha = 0.01,normalize=True)
LM1.fit(x , y)

# Coefficient values for all the independent variables
LM1.coef_
LM1.intercept_

# Getting Adjusted R squared value
LM1.score(x,y)

pred1 = LM1.predict(x)

# Rmse value
np.sqrt(np.mean((pred1-y)**2))

# Validating Model Assumptions
# Residuals Vs Fitted Values
plt.scatter(x = pred1 , y = y - pred1);plt.xlabel("Fitted");plt.ylabel("Residuals")
# Checking normal distribution 
plt.hist(pred1-y) # Normally distributed but skewed

# Predicted Vs Actual
plt.scatter(x = pred1 , y = y);plt.xlabel("Predicted");plt.ylabel("Actual")
plt.bar(height = pd.Series(LM1.coef_),x = x.columns) #clearly the coeff magnitudes vary by large extent

# Running a Ridge Regressor of set of alpha values and observing how the R-Squared
rmse = []
R_sqrd = []
alphas = np.arange(0,30,0.05)
for i in alphas:
    LM = Lasso(alpha = i,normalize=True)
    LM.fit(x , y)
    R_sqrd.append(LM.score(x , y))
    rmse.append(np.sqrt(np.mean((LM.predict(x) - y)**2)))
    
    
# Plotting RMSE,R_Squared values with respect to alpha values

# Alpha vs R_Squared values
plt.scatter(x=alphas,y=R_sqrd);plt.xlabel("alpha");plt.ylabel("R_Squared")

# Alpha vs RMSE
plt.scatter(x=alphas,y=rmse);plt.xlabel("alpha");plt.ylabel("rmse")

# R Squared value is constant for different alpha values
# The RMSE Value is lower for small alpha values - So we would use simple multiple regression


import statsmodels.formula.api as smf
formula = "Price" + "~" + "+".join(x.columns)
model = smf.ols(formula ,data=corolla).fit() # regression model

# Summary
model.summary()


# influence index plots
import statsmodels.api as sm
fig, ax = plt.subplots(figsize=(12,8))
fig = sm.graphics.influence_plot(model, alpha = 0.05 , ax = ax , criterion = "cooks") # Influential Points - 80

new_corolla = corolla.drop(corolla.index[[80]] , axis=0)
model2 = smf.ols(formula ,data=new_corolla).fit() # regression model

# Summary
model2.summary()
# Influce index plots
fig, ax = plt.subplots(figsize=(12,8))
fig = sm.graphics.influence_plot(model2, alpha = 0.05 , ax = ax , criterion = "cooks") # Points 960,221 seems to be outliers

#RMSE
pred2 = model2.predict(new_corolla)
res = pred2 - new_corolla.Price
rmse = np.sqrt(np.mean(res * res))
rmse

new_corolla = corolla.drop(corolla.index[[80 , 221 , 960]] , axis=0)
model3 = smf.ols(formula ,data=new_corolla).fit() # regression model

# Summary
model3.summary()
# Influce index plots
fig, ax = plt.subplots(figsize=(12,8))
fig = sm.graphics.influence_plot(model3, alpha = 0.05 , ax = ax , criterion = "cooks") # Points 960,221 seems to be outliers

#RMSE
pred3 = model3.predict(new_corolla)
res = pred3 - new_corolla.Price
rmse = np.sqrt(np.mean(res * res))
rmse

# Validating Model Assumptions
# Residuals Vs Fitted Values
plt.scatter(x = pred3 , y = new_corolla.Price - pred3);plt.xlabel("Fitted");plt.ylabel("Residuals")
# Checking normal distribution 
plt.hist(pred3 - new_corolla.Price) # Normally distributed but skewed

# Predicted Vs Actual
plt.scatter(x = pred3 , y = new_corolla.Price);plt.xlabel("Predicted");plt.ylabel("Actual")


# QQ plot for residuals 
import pylab          
import scipy.stats as st

# Checking Residuals are normally distributed
st.probplot(model3.resid_pearson, dist="norm", plot=pylab)


############ Homoscedasticity #######

# Residuals VS Fitted Values 
plt.scatter(pred3,model3.resid_pearson,c="r"),plt.axhline(y=0,color='blue');plt.xlabel("fitted_values");plt.ylabel("residuals")

# Spliting the data into train and test 
from sklearn.model_selection import train_test_split
train,test = train_test_split(new_corolla,test_size=0.2)


# preparing the model on train data 

model_train = smf.ols(formula,data=train).fit()

# train_data prediction
train_pred = model_train.predict(train)

# train residual values 
train_resid  = train_pred - train.Price

# RMSE value for train data 
train_rmse = np.sqrt(np.mean(train_resid*train_resid))

# prediction on test data set 
test_pred = model_train.predict(test)

# test residual values 
test_resid  = test_pred - test.Price

# RMSE value for test data 
test_rmse = np.sqrt(np.mean(test_resid*test_resid))

# Both test and train RMSE are close by
